File: functions/src/triggers/weekResetTrigger.ts
import * as functions from 'firebase-functions';
import { getFirestore } from 'firebase-admin/firestore';
import { WeekResetService } from '../services/weekResetService';
import * as logger from 'firebase-functions/logger';

/**
 * HTTP-callable Cloud Function that executes week_reset() for a given flat.
 *
 * Triggered by the grace-period Cloud Scheduler job:
 *   - Fires `grace_period_hours` after the latest task due_date_time in the current week
 *   - The scheduler sets the flatId via the request body
 *
 * In production, only the scheduler (or an admin) should call this.
 * The function is idempotent: re-running after a crash produces the same result
 * because it always reads current Firestore state inside a transaction.
 */
export const weekResetCallable = functions.https.onCall(
  async (request) => {
    const flatId = request.data?.flatId as string | undefined;
    if (!flatId) {
      throw new functions.https.HttpsError('invalid-argument', 'flatId is required');
    }

    const db = getFirestore();
    const service = new WeekResetService(db);
    await service.weekReset(flatId);
    return { success: true };
  },
);

/**
 * HTTP trigger variant used by Cloud Scheduler (which cannot call HTTPS callables directly).
 * Expects a JSON body: { "flatId": "<id>" }
 */
export const weekResetHttp = functions.https.onRequest(async (req, res) => {
  const flatId = req.body?.flatId as string | undefined;
  if (!flatId) {
    res.status(400).send({ error: 'flatId is required' });
    return;
  }

  try {
    const db = getFirestore();
    const service = new WeekResetService(db);
    await service.weekReset(flatId);
    res.status(200).send({ success: true });
  } catch (error) {
    logger.error('weekResetHttp failed', { flatId, error });
    res.status(500).send({ error: 'Internal error' });
  }
});

File: functions_python/triggers/week_reset_trigger.py
"""Cloud Functions triggers for weekly task rotation reset.

Triggered by Cloud Scheduler after the grace period expires.
Expects JSON body / callable data: {"flatId": "<id>"}
"""

from __future__ import annotations

import logging
from typing import Any

from firebase_functions import https_fn
from flask import Request, Response
from google.cloud import firestore

from services.week_reset_service import WeekResetService

logger = logging.getLogger(__name__)


@https_fn.on_call()  # type: ignore[untyped-decorator]
def week_reset_callable(req: https_fn.CallableRequest[Any]) -> dict[str, Any]:
    """HTTP-callable Cloud Function that executes week_reset() for a given flat."""
    flat_id: str = (req.data or {}).get("flatId", "")
    if not flat_id:
        raise https_fn.HttpsError(
            code=https_fn.FunctionsErrorCode.INVALID_ARGUMENT,
            message="flatId is required",
        )
    db = firestore.Client()
    WeekResetService(db).week_reset(flat_id)
    return {"success": True}


@https_fn.on_request()  # type: ignore[untyped-decorator]
def week_reset_http(req: Request) -> Response:
    """HTTP trigger variant used by Cloud Scheduler.

    Expects JSON body: {"flatId": "<id>"}
    """
    body = req.get_json(silent=True) or {}
    flat_id: str = body.get("flatId", "")
    if not flat_id:
        return Response({"error": "flatId is required"}, status=400, mimetype="application/json")

    try:
        db = firestore.Client()
        WeekResetService(db).week_reset(flat_id)
        return Response({"success": True}, status=200, mimetype="application/json")
    except Exception as exc:
        logger.error("week_reset_http failed flat_id=%s error=%s", flat_id, exc)
        return Response({"error": "Internal error"}, status=500, mimetype="application/json")
